### ISSUE TITLE
DrawTriangles/AppendVerticesAndIndicesForFilling doesn't always draw images with EvenOdd fill type

### ISSUE THREAD

**Issue by steampoweredtaco** [hajimehoshi/ebiten#2138]
Thanks for an awesome project on GO.  I found an issue when trying to call DrawTriangles ~100 times in a single draw tick.  The attached gist attached is a fairly basic program that demonstrates the issue and the source of a demonstration of the bug.  

You can see that some of the arcs I created with Arc and AppendVerticesAndIndiciesForFilling will flicker out and then back in or vice versa. In the example, as is, the flip flop only happens only once but there will always be some always missing from the location they should be at. 

The red circles are using DrawImage which should be next to the FilledCircles are used to more easily see where the filled circles are missing when they go missing. DrawImage doesn't suffer from this issue.

If I randomize the movements of each circle away from the current position in update typically the same filled circles will not show back up and sometimes you can see the filled circles somewhere else without the debug red circle.

Earlier today I could actually see in my main program with the debug circles the filled circles appearing in places without the debug circles hinting that they are being drawn but in the wrong locations.

Although the documentation says you must use EvenOdd fill type for DrawTriangles, if I use AlwaysFill the issue seems to go away. Is this an OK workaround or will it break something else?

Simple example:
https://gist.github.com/steampoweredtaco/23ff0600b95e3e7e49064f511f808683
![demoOfBug](https://user-images.githubusercontent.com/86767623/173258885-ce08aa27-1cec-489e-a8eb-284fb5e4018a.gif)

**Comment by hajimehoshi** [comment]
I'll take a look later, thank you for reporting!

**Comment by hajimehoshi** [comment]
I couldn't reproduce this issue on macOS. Which environment are you using? Thanks,

**Comment by steampoweredtaco** [comment]
Sorry about not mentioning that, I'm on windows 11 using an NVidia RTX 3090.

**Comment by hajimehoshi** [comment]
The DebugView says:

```
[3424] D3D12 ERROR: ID3D12CommandList::DrawIndexedInstanced: The current primitive topology in the Command List is D3D_PRIMITIVE_TOPOLOGY_UNDEFINED. This is invalid since the command list topology must match the Pipeline State topology type which cannot be undefined. [ EXECUTION ERROR #219: COMMAND_LIST_DRAW_INVALID_PRIMITIVETOPOLOGY]
[3424] D3D12 WARNING: ID3D12CommandList::DrawIndexedInstanced: An Index Buffer is expected, but none is set on the command list. This is OK, as reading from a missing Index Buffer is defined to return 0.However the developer probably did not intend to make use of this behavior.  [ EXECUTION WARNING #211: COMMAND_LIST_DRAW_INDEX_BUFFER_NOT_SET]
[3424] D3D12 ERROR: ID3D12CommandList::DrawIndexedInstanced: The depth stencil format does not match that specified by the current pipeline state. A null depth stencil view may only be bound when the pipeline state depth stencil format is UNKNOWN. (pipeline state = D24_UNORM_S8_UINT, DSV ID3D12Resource* = 0x0000000000000000:'(nullptr)') [ EXECUTION ERROR #615: DEPTH_STENCIL_FORMAT_MISMATCH_PIPELINE_STATE]
[3424] D3D12: Removing Device.
[3424] D3D12 MESSAGE: ID3D12Device::RemoveDevice: Device removal has been triggered for the following reason (DXGI_ERROR_DEVICE_REMOVED: Plug-and-play stop has been exercised on this adapter. Odds are good that the adapter might no longer be available). [ EXECUTION MESSAGE #234: DEVICE_REMOVAL_PROCESS_NOT_AT_FAULT]
```

**Comment by hajimehoshi** [comment]
Minimized test case to reproduce the warning (rather than blinking)

```
package main

import (
	"image"
	"image/color"
	"math"
	"math/rand"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/text"
	"github.com/hajimehoshi/ebiten/v2/vector"
	"golang.org/x/image/colornames"
	"golang.org/x/image/font"
	"golang.org/x/image/font/gofont/goregular"
	"golang.org/x/image/font/opentype"
)

type GameConfig struct {
	Width  int
	Height int
	Title  string
}

var (
	CFG = GameConfig{
		Width:  800,
		Height: 600,
		Title:  "Bugs?",
	}
)

var (
	emptyImage        = ebiten.NewImage(3, 3)
	debugCircleImage  *ebiten.Image
	emptyTextureImage = emptyImage.SubImage(image.Rect(1, 1, 2, 2)).(*ebiten.Image)
	face              font.Face
)

func NormalizeInt(v, fromMin, fromMax, toMin, toMax int) int {
	ratio := float64(toMax-toMin) / float64(fromMax-fromMin)
	return int(math.Round(float64(toMin) + ratio*float64(v-fromMin)))
}

func init() {
	emptyImage.Fill(color.White)
	img := image.NewRGBA(image.Rect(0, 0, 20, 20))
	debugCircleImage = ebiten.NewImageFromImage(img)
	emptyImage.Fill(colornames.Aquamarine)
	f, _ := opentype.Parse(goregular.TTF)
	face, _ = opentype.NewFace(f, &opentype.FaceOptions{
		Size: 12,
		DPI:  72,
	})
}

type Location struct {
	X int
	Y int
}

type Circle struct {
	id     int
	size   int
	color  color.RGBA
	center Location
}

type Game struct {
	screenWidth, screenHeight int
	circle                   Circle
}

func (g *Game) Update() error {
	return nil
}

func (g *Game) Draw(screen *ebiten.Image) {
	c := g.circle
	p := vector.Path{}
	p.Arc(float32(c.center.X), float32(c.center.Y), float32(c.size)/2, 0, 2*math.Pi, vector.Clockwise)
	filling, indicies := p.AppendVerticesAndIndicesForFilling(nil, nil)

	for i := range filling {
		filling[i].ColorR = float32(c.color.R) / 0xff
		filling[i].ColorG = float32(c.color.G) / 0xff
		filling[i].ColorB = float32(c.color.B) / 0xff
	}

	screen.DrawImage(debugCircleImage, nil)
	text.Draw(screen, "014678.,", face, c.center.X, c.center.Y, colornames.White)

	op2 := &ebiten.DrawTrianglesOptions{
		FillRule: ebiten.EvenOdd,
	}
	screen.DrawTriangles(filling, indicies, emptyTextureImage, op2)
}

func (g *Game) Layout(outsideWidth, outsideHeight int) (screenWidth, screenHeight int) {
	screenWidth = g.screenWidth
	screenHeight = g.screenHeight
	return
}

func NewGame(width, height int) *Game {
	g := &Game{
		screenWidth:  width,
		screenHeight: height,
		circle:       NewCircle(1, width, height),
	}
	return g
}

func NewCircle(id, maxX, maxY int) Circle {
	o := Circle{
		id:     id,
		size:   12,
		color:  colornames.Azure,
		center: Location{rand.Intn(maxX), rand.Intn(maxY)},
	}
	return o
}

func main() {
	ebiten.SetWindowSize(CFG.Width, CFG.Height)
	ebiten.SetWindowTitle(CFG.Title)
	g := NewGame(CFG.Width, CFG.Height)
	if err := ebiten.RunGame(g); err != nil {
		panic(err)
	}

}
```

**Comment by hajimehoshi** [comment]
Another warnings after some fixes:

```
[5344] D3D12 ERROR: ID3D12CommandList::DrawIndexedInstanced: The depth stencil format does not match that specified by the current pipeline state. A null depth stencil view may only be bound when the pipeline state depth stencil format is UNKNOWN. (pipeline state = D24_UNORM_S8_UINT, DSV ID3D12Resource* = 0x0000000000000000:'(nullptr)') [ EXECUTION ERROR #615: DEPTH_STENCIL_FORMAT_MISMATCH_PIPELINE_STATE]
```

**Comment by hajimehoshi** [comment]
I don't think this was fixed (and actually I've seen the same issue), but could you take a look at `2a5b64bf4d5804ebbb6e131d724ea39fc05190e0`?  Thanks.

```sh
go get github.com/hajimehoshi/ebiten/v2@2a5b64bf4d5804ebbb6e131d724ea39fc05190e0 
```

**Comment by steampoweredtaco** [comment]
Thanks for all the effort you've put into this. Would a workaround for now be to force opengl instead of directx on windows? I haven't looked to see if this is an option yet.

I'll check out the commit indicated, but it might be a week or so before I am able to.

**Comment by hajimehoshi** [comment]
> Thanks for all the effort you've put into this. Would a workaround for now be to force opengl instead of directx on windows? I haven't looked to see if this is an option yet.

You can do by calling `os.Setenv("EBITEN_GRAPHICS_LIBRARY", "opengl")` before the main loop starts.

**Comment by hajimehoshi** [comment]
Note to myself, these seems something wrong with constant buffer usages.

```diff
diff --git a/internal/graphicsdriver/directx/graphics_windows.go b/internal/graphicsdriver/directx/graphics_windows.go
index ee40174e9..b3c05e8eb 100644
--- a/internal/graphicsdriver/directx/graphics_windows.go
+++ b/internal/graphicsdriver/directx/graphics_windows.go
@@ -1254,6 +1254,7 @@ func (g *Graphics) drawTriangles(pipelineState *_ID3D12PipelineState, srcs [grap
        // Release constant buffers when too many ones were created.
        // This is needed espciallly for testings, where present is always false.
        if len(g.pipelineStates.constantBuffers[g.frameIndex]) >= 16 {
+               println("foo")
                if err := g.flushCommandList(g.drawCommandList); err != nil {
                        return err
                }
```

I guess the issue happens when the println reaches.

**Comment by hajimehoshi** [comment]
So, this happens when the draw commands are too many. With `-tags=ebitendebug`, there are over 300 commands for one frame:

```
Graphics commands:                                                                                                                                                                                                                                                                    
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 6, colorm: {}, mode: copy, filter: nearest, address: unsafe, even-odd: false                                                                                
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 126, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 126, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 126, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false                                                                       
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 33, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: true                                                                         
  draw-triangles: dst: 4 <- src: [1, (nil), (nil), (nil)], dst region: (x:0, y:0, width:800, height:600), num of indices: 120, colorm: {}, mode: source-over, filter: nearest, address: unsafe, even-odd: false   
...
```

**Comment by hajimehoshi** [comment]
@steampoweredtaco Hi, I think this was fixed in the latest Ebitengine (` 08c8c8089ad1b8dc5e3bf9909135ade54af0d124`).

By the way, in general, switching the even-odd property increased draw calls. In your case, the following code would be much more efficient:

```go
func (g *Game) Draw(screen *ebiten.Image) {
        for _, c := range g.circles {
                op := &ebiten.DrawImageOptions{
                        GeoM:          ebiten.GeoM{},
                        ColorM:        ebiten.ColorM{},
                        CompositeMode: 0,
                        Filter:        0,
                }
                op.GeoM.Translate(float64(c.center.X), float64(c.center.Y))
                screen.DrawImage(debugCircleImage, op)
                cord := fmt.Sprintf("%d, %.2f, %.2f, %.2f", c.id, float32(c.center.X), float32(c.center.Y), float32(c.size)/2)
                text.Draw(screen, cord, face, c.center.X, c.center.Y, colornames.White)
        }

        for _, c := range g.circles {
                p := vector.Path{}
                p.Arc(float32(c.center.X), float32(c.center.Y), float32(c.size)/2, 0, 2*math.Pi, vector.Clockwise)
                filling, indicies := p.AppendVerticesAndIndicesForFilling(nil, nil)
                for i := range filling {
                        filling[i].ColorR = float32(c.color.R) / 0xff
                        filling[i].ColorG = float32(c.color.G) / 0xff
                        filling[i].ColorB = float32(c.color.B) / 0xff
                }
                op := &ebiten.DrawTrianglesOptions{
                        FillRule: ebiten.EvenOdd,
                }
                screen.DrawTriangles(filling, indicies, emptyTextureImage, op)
        }
}
```
